save-common/src/commonMain/kotlin/org/cqfn/save/core/utils/CmdExecutorBase.kt

```diff
@@ -0,0 +1,103 @@
+package org.cqfn.save.core.utils
+
+import org.cqfn.save.core.config.TestConfig
+import org.cqfn.save.core.logging.logDebug
+import org.cqfn.save.core.plugin.ExtraFlags
+import org.cqfn.save.core.plugin.ExtraFlagsExtractor
+import org.cqfn.save.core.plugin.resolvePlaceholdersFrom
+
+import okio.Path
+import okio.Path.Companion.toPath
+
+/**
+ * FixMe: All plugins should use it for executing of commands in the future
+ *
+ * @property execCmd - execution command. Will be modified with extra flags and test files
+ */
+abstract class CmdExecutorBase(
+    val execCmd: String,
+    private val timeOutMillis: Long,
+    private val copyPaths: List<Path>,
+    private val extraFlagsExtractor: ExtraFlagsExtractor,
+    private val pb: ProcessBuilder,
+    private val testConfig: TestConfig,
+) {
+    private lateinit var constructedExecCmd: String
+
+    /**
+     * Simple method for extracting of extra flags, that will be used later in execCmd
+     */
+    private fun extractExtraFlags(): ExtraFlags {
+        val extraFlagsList = copyPaths.mapNotNull { extraFlagsExtractor.extractExtraFlagsFrom(it) }.distinct()
+        require(extraFlagsList.size <= 1) {
+            "Extra flags for all files in a batch should be same, but you have" +
+                    " ${extraFlagsList.size} different sets of flags inside it, namely $extraFlagsList"
+        }
+
+        return extraFlagsList.singleOrNull() ?: ExtraFlags("", "")
+    }
+
+    /**
+     * @param redirectTo a file where process output and errors should be redirected.
+     * If null, output will be returned as [ExecutionResult.stdout] and [ExecutionResult.stderr].
+     * @return the result after the execution of [constructedExecCmd]
+     */
+    fun executeCommandAndGetTestResults(redirectTo: Path?): ExecutionResult {
+        val time = timeOutMillis.times(copyPaths.size)
+        val execResult = pb.exec(constructedExecCmd, testConfig.getRootConfig().directory.toString(), redirectTo, time)
+
+        return ExecutionResult(execResult.code, execResult.getStdout(), execResult.stderr)
+    }
+
+    /**
+     * @param tmpDirName - the name of test file in the temporary directory
+     * @return constructed command for the execution. At the same time updates [constructedExecCmd]
+     */
+    fun constructExecCmd(tmpDirName: String): String {
+        val execFlagsAdjusted = resolvePlaceholdersFrom(execFlags(), extractExtraFlags(), constructFileNamesForExecCmd(tmpDirName))
+        constructedExecCmd = "$execCmd $execFlagsAdjusted"
+        return constructedExecCmd
+    }
+
+    private fun constructFileNamesForExecCmd(tmpDirName: String): String {
+        // joining test files to string with a batchSeparator if the tested tool supports processing of file batches
+        // NOTE: SAVE will pass relative paths of Tests (calculated from testRootConfig dir) into the executed tool
+        val fileNamesForExecCmd = when (wildCardInDirectoryMode()) {
```

Comment:
Just for interest, why do you prefer `when` here instead of `.let` construction?
We can have only two conditions: `null` and `not null`. And `let` copes with this task well, also here  you have double invocation of `wildCardInDirectoryMode()`, previously we used `it`

Reply:
fixed